### PR TITLE
ADR to resolve model dependency downstream integration

### LINKED ISSUE #23: ADR for failure prediction model deployment
As a developer/maintainer of machine learning models in ceph, I want to understand the python package related issues and constraints in integrating these models in ceph.

**Acceptance Criterion**
- [ ] ADR describing the current scikit-learn issue in detail, and what potential solutions could look like

### PR DESCRIPTION
**Pull request by chauhankaranraj**
Fixes #23 

Hey team, this PR adds a first draft of the ADR to describing `scikit-learn` downstream packaging issue, as well as some potential solutions and their pros/cons. This list may not be exhaustive, so please feel free to add any other potential solutions I missed :)

Let's have a discussion on this PR about what the best path forward is, and then I can update the ADR to reflect our conversation and decision.

### CODE REVIEW

**Comment by durandom** [comment]
Good outline, I have no objections :) LGTM

File: docs/adr/0001-handling-dependencies-downstream.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+* Con: Can turn into essentially "copying" model implementations from other libraries, which does not seem like a good practice.
+
+### Option 5
+
+* Pro: Eliminates dependence on external packages.
+* Pro: Provides flexibility in what pacakges can be used.
+* Pro: Provides opportunity to use Thoth recommended software stack to maximize security and performance.
+* Pro: Ceph already had a module ([diskprediction_cloud](https://github.com/ceph/ceph/tree/nautilus/src/pybind/mgr/diskprediction_cloud)) designed to query an external service for failure predictions. But this had to be removed because the external service dropped support. Under this option, this module can be revived.
+* Con: Disk failure prediction will not work "out of the box"; it would depend on the deployed service.
+* Con: Some users may be reluctant to send their disk health data to the service.
```
Comment (liewegas) [inline docs/adr/0001-handling-dependencies-downstream.md:57]: Con: model will not be usable in a disconnected environment

File: docs/adr/0001-handling-dependencies-downstream.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+* How would the solution affect user experience?
+* How difficult would it be to maintain the solution, long term?
+* What would need to happen if we want to use a different pacakge?
+* What would need to happen if we want to increase model complexity?
+* What would need to happen if we want to add, remove or replace models?
+
+## Decision Outcome [to be discussed]
+
+Suggested choice: Option #1 for current models, Option #5 for future models
```
Comment (liewegas) [inline docs/adr/0001-handling-dependencies-downstream.md:69]: I think option 5 is something we should eventually pursue, but it cannot replace the ability to use the models in a disconnected environment.  Option 1 is the obvious immediate choice.

I would suggest a wait-and-see strategy for options 2-3 and let the engineering teams do whatever is necessary or most expedient to ship the software.

Option 4 is not worth considering IMO; packaging overhead shouldn't prevent us from using existing FOSS software (if it does I'm not sure I can explain what Red Hat's business model is!)

File: docs/adr/0001-handling-dependencies-downstream.md (hunk 1 above)
Comment (chauhankaranraj) [reply docs/adr/0001-handling-dependencies-downstream.md:57]: Yep, that's the idea I was trying to capture in the first `Con`, but I'll just explicitly add this as a separate bullet to make it super clear :)

File: docs/adr/0001-handling-dependencies-downstream.md (hunk 2 above)
Comment (chauhankaranraj) [reply docs/adr/0001-handling-dependencies-downstream.md:69]: > I think option 5 is something we should eventually pursue, but it cannot replace the ability to use the models in a disconnected environment. Option 1 is the obvious immediate choice.

Yeah that makes sense.

> Option 4 is not worth considering IMO; packaging overhead shouldn't prevent us from using existing FOSS software (if it does I'm not sure I can explain what Red Hat's business model is!)

100% agreed, I had just put it in the doc to make sure we've covered _all_ possible options, including good ones and not-so-good ones :)

File: docs/adr/0001-handling-dependencies-downstream.md (hunk 1 above)
Comment (goern) [reply docs/adr/0001-handling-dependencies-downstream.md:57]: I think the operational challenges of running a containerized service in an air-gapped environment are not specific to this ADR: we need to have some procedure to get the container image into the env, then we run the service. any models we ship to customers would anyway be in container images, or would we repackaged them as RPM?!

**Comment by goern** [comment]
@chauhankaranraj for providing such a service in upstream ceph, what needs to be done? 

and if we got that implemented, and build if via AICoE-CI and publish it to quay, 

@liewegas  what are the steps to put it in the downstream product?

**Comment by chauhankaranraj** [comment]
@goern @liewegas hey guys, I updated the outcome/next steps based on our discussion. Could you ptal and lmk if anything seems inaccurate?

**Comment by chauhankaranraj** [comment]
hey @liewegas, if the next steps look good to you could you please give an lgtm :smile: (or ofc suggest changes)?

**Review by liewegas — APPROVED** [review]

**Pull request merged**